### PR TITLE
Using a simpler ViewIndex instead of MultiIndex in VariableView

### LINKED ISSUE #224: Refactor internals of VariableView
`VariableView` (not to be confused with `VariableSlice`) is currently using `MultiIndex` internally. This was simply for convenience during the prototype phase.

- Implement an equivalent to `MultiIndex`, supporting iteration only for 1 array.
- Make sure it is as fast or faster.
- Maybe add some benchmarks. If it is sufficiently fast we might be able to use `VariableView` everywhere, instead of `scipp:span`. This would reduce the code complexity when implementing operations with `VariableConceptT<T>`.

### PR DESCRIPTION
**Pull request by nvaytet**
`MultiIndex` has up to 4 sub-indices which can be iterated at the same time.
This is not necessary for `VariableView`.
A simpler `ViewIndex`, based on `MultiIndex` was created, which still offers support for broadcasting and transposing, as well as slicing.
This first pass was simply to remove the unnecessary parts inside the `MultiIndex` class, with the hope that it would improve performance.

This is the output from `google-benchmark`:

**Using `ViewIndex`:**
```
Running benchmark/variable_view_benchmark
Run on (24 X 3100 MHz CPU s)
CPU Caches:
  L1 Data 32K (x12)
  L1 Instruction 32K (x12)
  L2 Unified 256K (x12)
  L3 Unified 15360K (x2)
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
---------------------------------------------------------
Benchmark                  Time           CPU Iterations
---------------------------------------------------------
BM_ViewIndex/4          9475 ns       9474 ns      73678   805.258M items/s
BM_ViewIndex/8         18824 ns      18823 ns      37177   810.629M items/s
BM_ViewIndex/16        37817 ns      37816 ns      18388   807.003M items/s
BM_ViewIndex/32        75835 ns      75833 ns       9228   804.868M items/s
BM_ViewIndex/64       151650 ns     151648 ns       4613   804.956M items/s
BM_ViewIndex/128      304204 ns     304202 ns       2307   802.561M items/s
BM_ViewIndex/256      610057 ns     610038 ns       1147   800.412M items/s
BM_ViewIndex/512     1245476 ns    1245440 ns        571    784.11M items/s
BM_ViewIndex/1024    2741253 ns    2741185 ns        255   712.511M items/s
BM_ViewIndex/2048    5637325 ns    5637186 ns        124   692.943M items/s
BM_ViewIndex/4096   11278059 ns   11277769 ns         62   692.735M items/s
BM_ViewIndex/8192   23030247 ns   23029656 ns         31   678.473M items/s
```

**Using `MultiIndex`:**
```
Running benchmark/variable_view_benchmark
Run on (24 X 3100 MHz CPU s)
CPU Caches:
  L1 Data 32K (x12)
  L1 Instruction 32K (x12)
  L2 Unified 256K (x12)
  L3 Unified 15360K (x2)
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
---------------------------------------------------------
Benchmark                  Time           CPU Iterations
---------------------------------------------------------
BM_ViewIndex/4          9641 ns       9641 ns      72475   791.379M items/s
BM_ViewIndex/8         19004 ns      19004 ns      36855   802.915M items/s
BM_ViewIndex/16        38079 ns      38079 ns      18385   801.434M items/s
BM_ViewIndex/32        76869 ns      76866 ns       9102   794.041M items/s
BM_ViewIndex/64       152986 ns     152982 ns       4576   797.939M items/s
BM_ViewIndex/128      305848 ns     305842 ns       2289   798.257M items/s
BM_ViewIndex/256      616002 ns     615939 ns       1137   792.743M items/s
BM_ViewIndex/512     1270649 ns    1270638 ns        554    768.56M items/s
BM_ViewIndex/1024    2944054 ns    2944026 ns        237    663.42M items/s
BM_ViewIndex/2048    5946242 ns    5946170 ns        118   656.935M items/s
BM_ViewIndex/4096   11494767 ns   11494488 ns         61   679.674M items/s
BM_ViewIndex/8192   23025756 ns   23025518 ns         30   678.595M items/s
```

The size of the arrays are:
```
4    * 2000 * 8 = 64000     bytes
8    * 2000 * 8 = 120000    bytes
16   * 2000 * 8 = 256000    bytes
32   * 2000 * 8 = 512000    bytes
64   * 2000 * 8 = 1024000   bytes
128  * 2000 * 8 = 2048000   bytes
256  * 2000 * 8 = 4096000   bytes
512  * 2000 * 8 = 8192000   bytes
1024 * 2000 * 8 = 16384000  bytes
2048 * 2000 * 8 = 32768000  bytes
4096 * 2000 * 8 = 65536000  bytes
8192 * 2000 * 8 = 131072000 bytes
```

The benchmark is performing a sum of doubles, so the speeds are of the order of 6.4G/s

**Further note**

We investigated different containers for the `m_delta`, `m_coord` & `m_extent` arrays, and their impact on peformance when iterating over a variable view.
Using `std::array` or `C-style arrays` give good performance (7.5 Gb/s) **as long as a range based loop is used**:
```C++
for ( const auto x : view ) {
  sum += x;
}
```
If a loop which explicitly accesses the `begin()` and `end()` of the container is used, e.g.
```C++
for ( auto it = view.begin(); it != view.end(); ++it ) {
  sum += *it;
}
```
then the results differ widely:
- using `std::array` is 80x slower than above, at ~90 Mb/s
- using `C-style arrays` is 20x slower than above, at ~330 Mb/s

We can recover the maximum performance by storing the `view.end()` in a variable, e.g.
```C++
auto iend = view.end();
for ( auto it = view.begin(); it != iend; ++it ) {
  sum += *it;
}
```
for both `std::array` and `C-style arrays`.

Finally, when using `C-style arrays`, we get a compilation warning with the `gcc` compiler:
```
In file included from /home/nvaytet/work/code/scipp/source/core/./variable_view.h:14:0,
                 from /home/nvaytet/work/code/scipp/source/core/./variable.h:18,
                 from /home/nvaytet/work/code/scipp/source/core/./transform.h:9,
                 from /home/nvaytet/work/code/scipp/source/core/test/transform_test.cpp:9:
/home/nvaytet/work/code/scipp/source/core/./view_index.h: In constructor ‘scipp::core::ViewIndex::ViewIndex(const scipp::core::Dimensions&, const scipp::core::Dimensions&)’:
/home/nvaytet/work/code/scipp/source/core/./view_index.h:37:22: warning: array subscript is above array bounds [-Warray-bounds]
           m_delta[d] -= m_delta[d2];
           ~~~~~~~~~~~^~~~~~~~~~~~~~
```
which disappears when switching to `std::array`. This warning is not given by the `clang` compiler, and is not fully understood as `d2` is always less than `d` and should never overflow the array bounds.

We decided to go with `std::array` as our final choice to avoid the warning, as the performance is identical to `C-style arrays`, as long as range based loops are used.

Fixes #224

### CODE REVIEW

File: core/dimensions.h [hunk 1]
```diff
@@ -16,6 +16,8 @@
 
 namespace scipp::core {
 
+#define NDIM_MAX 6
```
Comment (SimonHeybrock) [inline core/dimensions.h]: Since C++11, use `constexpr` instead of a define.

File: core/view_index.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      offset *= m_extent[d];
+    }
+    setIndex(0);
+  }
+
+  void increment() {
+    m_index += m_delta[0];
+    ++m_coord[0];
+    scipp::index d = 0;
+    while ((m_coord[d] == m_extent[d]) && (d < NDIM_MAX - 1)) {
```
Comment (SimonHeybrock) [inline core/view_index.h]: I'd be interested to see your benchmarks. In the original multi-index implementation it made a big difference how this was implemented. Having `increment` as a large function prevented inlining, making it really slow (I split it into 2 functions). Is `increment` getting inlined with your implementation?

**Comment by SimonHeybrock** [comment]
Regarding your benchmark: You should test with data that fits into L1 and data that fits into L2 to see the true performance.

File: benchmark/variable_view_benchmark.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  for (auto _ : state) {
+    double sum = 0.0;
+    for ( const auto x : view ) {
+      sum += x;
+    }
+    benchmark::DoNotOptimize(sum);
+  }
+
+  state.SetItemsProcessed(state.iterations() * count);
```
Comment (SimonHeybrock) [inline benchmark/variable_view_benchmark.cpp]: Maybe just use `view.size()`, so we avoid future "bugs" like the one we had with `counts` being derived from the wrong dimensions variable.

File: core/view_index.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      m_index += m_delta[d + 1];
+      ++m_coord[d + 1];
+      m_coord[d] = 0;
+      ++d;
+    }
```
Comment (SimonHeybrock) [inline core/view_index.h]: So in the end you did not move this to its own function? Is it still getting the same performance?

File: core/view_index.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  scipp::index get() const { return m_index; }
+  scipp::index index() const { return m_fullIndex; }
+
+  bool operator==(const ViewIndex &other) const {
+    return m_fullIndex == other.m_fullIndex;
+  }
+
+private:
+  scipp::index m_index{0};
+  scipp::index m_delta[NDIM_MAX] = {0, 0, 0, 0, 0, 0};
```
Comment (SimonHeybrock) [inline core/view_index.h]: Can you also add the comments we had in `MultiIndex`, explaining why this is not using `std::array`?

File: core/view_index.h (hunk 4 above)
Comment (nvaytet) [reply core/view_index.h]: Yes i checked that the performance was the same and it was still inlining everything in the assembly code.

File: core/view_index.h (hunk 5 above)
Comment (nvaytet) [reply core/view_index.h]: :+1:

File: core/view_index.h [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  // with the GCC compiler:
+  //
+  //  warning: array subscript is above array bounds [-Warray-bounds]
+  //
+  // which disappears when switching to std::array. This warning is not given
+  // by the CLANG compiler, and is not fully understood as d2 is always less
+  // than d and should never overflow the array bounds.
+  // We decided to go with std::array as our final choice to avoid the warning,
+  // as the performance is identical to C-style arrays, as long as range based
+  // loops are used.
```
Comment (SimonHeybrock) [inline core/view_index.h]: :+1: for the detailed comment.

**Pull request merged**